File: hoffman-agents-node/examples/05_self_ref_ablation/self_ref_ablation.js
/**
 * Self-Reference ON/OFF Contrast — The Cleanest Causal Result
 *
 * One parameter changes everything:
 *   ON (threshold=0.25): 100% agents lock "I"
 *   OFF (threshold=1.5):   0% agents ever lock
 *
 * Without the "I" attractor mechanism, agents don't develop identity.
 */
const { ConsciousAgent, WorldState, SelfTokenState, ExperienceSpace } = require('../../src/index');

function strangeLoopScore(agent) {
  const out = agent._lastOutput;
  const nI = out.filter(t => t === 'I').length;
  return nI === 0 ? 0 : Math.min(0.5 * nI, 2);
}

function runCondition(name, lockThreshold, nAgents = 8, nSteps = 500) {
  console.log(`\n  ── ${name} ──`);
  const agents = [];
  for (let i = 0; i < nAgents; i++) {
    const st = new SelfTokenState({ lockThreshold });
    const exp = new ExperienceSpace({ selfToken: st });
    agents.push(new ConsciousAgent({ agentId: `Agent_${String(i).padStart(2, '0')}`, experience: exp }));
  }

  const lockGenerations = new Set();
  const outputsLog = {};

  for (let step = 0; step < nSteps; step++) {
    for (let i = 0; i < agents.length; i++) {
      const ws = WorldState.fromSequence('world', [`state_${step % 10}`]);
      const out = agents[i].step(ws);
      if (!outputsLog[agents[i].agentId]) outputsLog[agents[i].agentId] = [];
      outputsLog[agents[i].agentId].push(out.sequenceStr);
      if (out.iLocked) lockGenerations.add(agents[i].agentId);
    }
  }

  const lockRate = lockGenerations.size / nAgents;
  const loopScores = [];
  for (const aid of Object.keys(outputsLog).sort()) {
    const recent = outputsLog[aid].slice(-100);
    const score = recent.reduce((s, seq) => s + strangeLoopScore({ _lastOutput: seq.split(' ') }), 0) / Math.max(recent.length, 1);
    loopScores.push(score);
  }
  const meanLoop = loopScores.reduce((a, b) => a + b, 0) / loopScores.length;

  let nonTrivial = 0, total = 0;
  for (const seqs of Object.values(outputsLog)) {
    for (const s of seqs.slice(-100)) { total++; if (s !== 'wait') nonTrivial++; }
  }
  const outputSync = nonTrivial / Math.max(total, 1);

  console.log(`    Agents:          ${nAgents}`);
  console.log(`    Lock threshold:  ${lockThreshold}`);
  console.log(`    Lock rate:       ${(lockRate * 100).toFixed(0)}%`);
  console.log(`    Mean loop depth: ${meanLoop.toFixed(3)}`);
  console.log(`    Non-trivial out: ${(outputSync * 100).toFixed(1)}%`);

  return { lockRate, meanLoopDepth: meanLoop, outputSyncRatio: outputSync, locked: lockGenerations.size, unlocked: nAgents - lockGenerations.size };
}

function main() {
  const t0 = Date.now();
  console.log('='.repeat(62));
  console.log('Self-Reference ON/OFF Contrast');
  console.log('='.repeat(62));
  console.log('\nThe same world, same architecture — one parameter changes everything.\n');

  const on = runCondition('Self-Reference ON  (threshold=0.25)', 0.25);
  const off = runCondition('Self-Reference OFF (threshold=1.5)', 1.5);

  console.log(`\n  Done in ${((Date.now() - t0) / 1000).toFixed(1)}s\n`);
  console.log('='.repeat(62));
  console.log('Result Summary');
  console.log('='.repeat(62));
  console.log(`  Lock rate:        ON ${(on.lockRate*100).toFixed(0)}%  OFF ${(off.lockRate*100).toFixed(0)}%`);
  console.log(`  Mean loop depth:  ON ${on.meanLoopDepth.toFixed(3)}  OFF ${off.meanLoopDepth.toFixed(3)}`);
  console.log(`  Non-trivial out:  ON ${(on.outputSyncRatio*100).toFixed(0)}%  OFF ${(off.outputSyncRatio*100).toFixed(0)}%`);

  if (on.lockRate === 1 && off.lockRate === 0) {
    console.log('\n  ✓ Absolute contrast — self-reference is causally necessary');
  } else {
    console.log(`\n  ~ Partial contrast (${(on.lockRate*100).toFixed(0)}% vs ${(off.lockRate*100).toFixed(0)}%)`);
  }
}

main();

File: hoffman-agents-python/examples/05_self_ref_ablation/self_ref_ablation.py
"""
Self-Reference ON/OFF Contrast — The Cleanest Causal Result

In the same architecture, with the same world, same topology, same
everything — one parameter changes everything:

  Self-Reference ON:  lock_threshold=0.25  → 100% agents lock, ~gen 60
  Self-Reference OFF: lock_threshold=1.5   →   0% agents ever lock

Without the "I" attractor mechanism, agents do not develop stable
identities. This demonstrates that self-reference is causally
necessary for conscious-agent-like behavior.

Expected:
  ON:  attractor_lock_rate=1.0, mean_strange_loop≈1.0
  OFF: attractor_lock_rate=0.0, mean_strange_loop≈0.0
  Invariant to network size (N=2..32)
"""
from conscious_agent import ConsciousAgent, WorldState, SelfTokenState, ExperienceSpace
import numpy as np
import time


def strange_loop_score(tokens):
    """Simple self-reference score from token list."""
    n_i = sum(1 for t in tokens if t == "I")
    if n_i == 0:
        return 0.0
    return min(0.5 * n_i, 2.0)


def run_condition(name, lock_threshold, n_agents=8, n_steps=500):
    """Run agents in a shared world with given lock_threshold."""
    print(f"\n  ── {name} ──")
    agents = []
    for i in range(n_agents):
        st = SelfTokenState(lock_threshold=lock_threshold)
        exp = ExperienceSpace(self_token=st)
        agent = ConsciousAgent(agent_id=f"Agent_{i:02d}", experience=exp)
        agents.append(agent)

    lock_generations = []
    outputs_log = {f"Agent_{i:02d}": [] for i in range(n_agents)}

    for step in range(n_steps):
        for i, agent in enumerate(agents):
            ws = WorldState.from_sequence("world", [f"state_{step % 10}"])
            out = agent.step(ws)
            outputs_log[agent.agent_id].append(out.sequence_str)

            if out.i_locked and agent.agent_id not in lock_generations:
                lock_generations.append(agent.agent_id)

    # Metrics
    lock_rate = len(lock_generations) / n_agents
    loop_scores = []
    for aid in sorted(outputs_log.keys()):
        seq = outputs_log[aid]
        # Mean strange loop over last 100 outputs
        recent = seq[-100:]
        score = sum(strange_loop_score(s.split()) for s in recent) / max(len(recent), 1)
        loop_scores.append(score)
    mean_loop = np.mean(loop_scores)

    # Count non-trivial outputs
    non_trivial = 0
    total = 0
    for aid, seqs in outputs_log.items():
        for s in seqs[-100:]:
            total += 1
            if s != "wait":
                non_trivial += 1
    output_sync = non_trivial / max(total, 1)

    locked_list = [aid for aid in sorted(outputs_log.keys())
                   if aid in lock_generations]
    unlocked_list = [aid for aid in sorted(outputs_log.keys())
                     if aid not in lock_generations]

    print(f"    Agents:          {n_agents}")
    print(f"    Lock threshold:  {lock_threshold}")
    print(f"    Lock rate:       {lock_rate:.0%}")
    print(f"    Mean loop depth: {mean_loop:.3f}")
    print(f"    Non-trivial out: {output_sync:.1%}")
    if locked_list:
        print(f"    Locked:          {', '.join(locked_list)}")
    if unlocked_list:
        print(f"    Unlocked:        {', '.join(unlocked_list)}")

    return {
        "condition": name,
        "lock_rate": lock_rate,
        "mean_loop_depth": mean_loop,
        "output_sync_ratio": output_sync,
        "locked": len(lock_generations),
        "unlocked": n_agents - len(lock_generations),
        "non_trivial_output_pct": output_sync,
    }


def main():
    np.random.seed(42)
    t0 = time.time()

    print("=" * 62)
    print("Self-Reference ON/OFF Contrast")
    print("=" * 62)
    print()
    print("The same world, same topology, same agent architecture.")
    print("One parameter — lock_threshold — changes everything.\n")

    on = run_condition("Self-Reference ON  (threshold=0.25)", lock_threshold=0.25)
    off = run_condition("Self-Reference OFF (threshold=1.5)",  lock_threshold=1.5)

    elapsed = time.time() - t0
    print(f"\n  Completed in {elapsed:.1f}s\n")

    print("=" * 62)
    print("Result Summary")
    print("=" * 62)
    print(f"  {'Metric':<30s} {'ON':<15s} {'OFF':<15s}")
    print(f"  {'─'*58}")
    print(f"  {'Lock rate':<30s} {on['lock_rate']:.0%}          {off['lock_rate']:.0%}")
    print(f"  {'Mean loop depth':<30s} {on['mean_loop_depth']:<15.3f} {off['mean_loop_depth']:<15.3f}")
    print(f"  {'Non-trivial output':<30s} {on['output_sync_ratio']:.0%}          {off['output_sync_ratio']:.0%}")
    print(f"  {'Agents locked':<30s} {on['locked']:<15d} {off['unlocked']:<15d}")
    print()

    if on["lock_rate"] == 1.0 and off["lock_rate"] == 0.0:
        print("  ✓ RESULT: Absolute contrast — self-reference is causally")
        print("    necessary for stable identity formation.")
    elif on["lock_rate"] > off["lock_rate"]:
        print(f"  ~ RESULT: Partial contrast ({on['lock_rate']:.0%} vs {off['lock_rate']:.0%})")
    else:
        print(f"  ✗ Expected ON > OFF, got the inverse")

    print()
    print("  Reference: archive/hoffman_experiments/03_self_ref_ablation/")
    print("  Original result: ON 100% lock, OFF 0% lock (invariant to N=2..32)")


if __name__ == "__main__":
    main()
